atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java

```diff
@@ -0,0 +1,603 @@
+                            // Someone else changed the state on us, just start
+                            // over.
+                        }
+                    }
+                    break;
+
+                case ELEVATED:
+                    // Elevated state, but is it expired?
+                    if (System.currentTimeMillis() < stateLocal.elevatedStateTimeMillis + COOLDOWN_MILLISECONDS) {
+                        // Nope, the purest case -- delegate and hope for the
+                        // best (there is no twelve to which to turn it).
+                        return stateLocal.dataSourcePool.getConnection();
+                    } else {
+                        // Ah, but the elevated state is expired, we should amp
+                        // it back down.  But first, blah blah blah locks.
+                        synchronized (this) {
+                            if (state == stateLocal) {
+                                // The state hasn't changed on us, we can
+                                // perform the unelevation and start
+                                // getConnectionInternal() over again.
+                                state = normalState(state);
+                            } else {
+                                // Someone else changed the state on us, just
+                                // start over.
+                            }
+                        }
+                    }
+                    break;
+
+                case CLOSED:
+                    throw new SQLException("Hikari connection pool already closed!", stateLocal.closeTrace);
+            }
+
+            // fall throughs are spins
+        }
+    }
+
+    private void logConnectionFailure() {
+        log.error("Failed to get connection from the datasource "
+                + "db-pool-" + dbConfig.getConnId() + "-" + dbConfig.getDbLogin()
+                + ". Please check the jdbc url (" + (dbConfig.getUrl()) + ")"
+                + ", the password, and that the secure server key is correct for the hashed password.");
+    }
+
+    /**
+     * Test an initialized dataSourcePool by running a simple query.
+     */
+    private void testDataSource(HikariDataSource dataSourcePool) throws SQLException {
+        try {
+            Connection conn = null;
+
+            try {
+                conn = dataSourcePool.getConnection();
+                testConnection(conn);
+            } finally {
+                if (conn != null) {
+                    conn.close();
+                }
+            }
+        } catch (SQLException e) {
+            logConnectionFailure();
+            throw e;
+        }
+    }
+
+    private boolean testConnection(Connection conn) throws SQLException {
+        boolean isValid = false;
+        Statement stmt = conn.createStatement();
+        try {
+            ResultSet rs = stmt.executeQuery(dbConfig.getType().getTestQuery());
+            isValid = rs.next();
+            rs.close();
```

Comment:
this needs to be try-with-resources to ensure we close the result set cleanly if we get into a bad state.